app/forms/candidate_interface/science_gcse_grade_form.rb

```diff
@@ -0,0 +1,236 @@
+        physics: physics_grade,
+        chemistry: chemistry_grade,
+      }
+
+      qualification.update(grade: set_grade, grades: triple_award_grades, subject: subject)
+    end
+
+    def save_year
+      if valid?(:award_year)
+        qualification.update(grade: set_grade, award_year: award_year)
+        return true
+      end
+
+      false
+    end
+
+  private
+
+    def award_year_is_not_in_the_future
+      date_limit = Time.zone.now.year.to_i + 1
+      errors.add(:award_year, :in_future, date: date_limit) if award_year.to_i >= date_limit
+    end
+
+    def award_year_is_invalid
+      errors.add(:award_year, :invalid)
+    end
+
+    def award_year_is_a_valid_date
+      if valid_year?(award_year)
+        award_year_is_not_in_the_future
+      else
+        award_year_is_invalid
+      end
+    end
+
+    def validate_grade_format
+      return if qualification.qualification_type.nil? ||
+        qualification.qualification_type == 'other_uk' ||
+        qualification.qualification_type == 'non_uk' ||
+        grade.nil? ||
+        subject == ApplicationQualification::SCIENCE_TRIPLE_AWARD
+
+      qualification_rexp = invalid_grades[qualification.qualification_type.to_sym]
+
+      if qualification_rexp && grade.match(qualification_rexp)
+        if subject == ApplicationQualification::SCIENCE_SINGLE_AWARD
+          errors.add(:single_award_grade, :invalid)
+        elsif subject == ApplicationQualification::SCIENCE_DOUBLE_AWARD
+          errors.add(:double_award_grade, :invalid)
+        else
+          errors.add(:grade, :invalid)
+        end
+      end
+    end
+
+    def triple_award_grades_format
+      return unless triple_award?
+
+      qualification_rexp = invalid_grades[qualification.qualification_type.to_sym]
+
+      if qualification_rexp && biology_grade.match(qualification_rexp)
+        errors.add(:biology_grade, :invalid)
+      end
+
+      if qualification_rexp && chemistry_grade.match(qualification_rexp)
+        errors.add(:chemistry_grade, :invalid)
+      end
+
+      if qualification_rexp && physics_grade.match(qualification_rexp)
+        errors.add(:physics_grade, :invalid)
+      end
+    end
+
+    def grade_length
+      if grade.blank? && subject == ApplicationQualification::SCIENCE_SINGLE_AWARD
+        errors.add(:single_award_grade, :blank)
+      elsif grade.blank? && subject == ApplicationQualification::SCIENCE_DOUBLE_AWARD
+        errors.add(:double_award_grade, :blank)
+      elsif subject == ApplicationQualification::SCIENCE_TRIPLE_AWARD
+        if biology_grade.blank?
+          errors.add(:biology_grade, :blank)
+        end
+
+        if chemistry_grade.blank?
+          errors.add(:chemistry_grade, :blank)
+        end
+
+        if physics_grade.blank?
+          errors.add(:physics_grade, :blank)
+        end
+      elsif grade.blank?
+        errors.add(:grade, :blank)
+      end
+    end
+
+    def invalid_grades
+      {
+        gcse: /[^1-9A-GU\*\s\-]/i,
```

Comment:
Should we be using a set or enum for this instead? Especially if we move away from the autocomplete to text inputs.

I've tested locally and this would let through E*

Reply:
Yes that's a good point. I do think we need to check what the situation is with international GCSE equivalents and validation. It appears to use this regex as well. Even if we do have autocomplete, we will need validations in place for progressive enhancement (i.e no JS turned on)

Reply:
I think I prefer enum as it's a lot clearer what is and is not allowed :thinking:

Reply:
I've removed the regex check and now checking that the submitted grade is included in the values defined here - https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/config/initializers/gcse_grades.rb